patch_subset/cbor/cbor_utils.cc

```diff
@@ -0,0 +1,360 @@
+      // will eventually call cbor_decref(), the item has a new "owner".
+      cbor_item_t* value = pair->value;
+      if (value == nullptr) {
+        return StatusCode::kInvalidArgument;
+      }
+      out.reset(cbor_incref(value));
+      return StatusCode::kOk;
+    }
+  }
+  // Field not found.
+  out.reset(nullptr);
+  return StatusCode::kNotFound;
+}
+
+StatusCode CborUtils::GetUInt64Field(const cbor_item_t& map, int field_number,
+                                     optional<uint64_t>& out) {
+  cbor_item_unique_ptr field = empty_cbor_ptr();
+  StatusCode sc = GetField(map, field_number, field);
+  if (sc == StatusCode::kNotFound) {
+    out.reset();
+    return StatusCode::kOk;
+  } else if (sc != StatusCode::kOk) {
+    return sc;
+  }
+  uint64_t result;
+  sc = DecodeUInt64(*field, &result);
+  if (sc != StatusCode::kOk) {
+    return sc;
+  }
+  out.emplace(result);
+  return StatusCode::kOk;
+}
+
+// TODO: This method could ignore failures, returning "", if we want.
+StatusCode CborUtils::GetStringField(const cbor_item_t& map, int field_number,
+                                     optional<string>& out) {
+  cbor_item_unique_ptr field = empty_cbor_ptr();
+  StatusCode sc = GetField(map, field_number, field);
+  if (sc == StatusCode::kNotFound) {
+    out.reset();
+    return StatusCode::kOk;
+  } else if (sc != StatusCode::kOk) {
+    return sc;
+  }
+  string result;
+  sc = DecodeString(*field, result);
+  if (sc != StatusCode::kOk) {
+    return sc;
+  }
+  out.emplace(result);
+  return StatusCode::kOk;
+}
+
+// TODO: This method could ignore failures, returning "", if we want.
+StatusCode CborUtils::GetBytesField(const cbor_item_t& map, int field_number,
+                                    optional<string>& out) {
+  cbor_item_unique_ptr field = empty_cbor_ptr();
+  StatusCode sc = GetField(map, field_number, field);
+  if (sc == StatusCode::kNotFound) {
+    out.reset();
+    return StatusCode::kOk;
+  } else if (sc != StatusCode::kOk) {
+    return sc;
+  }
+  string result;
+  sc = DecodeBytes(*field, result);
+  if (sc != StatusCode::kOk) {
+    return sc;
+  }
+  out.emplace(result);
+  return StatusCode::kOk;
+}
+
+StatusCode CborUtils::GetProtocolVersionField(const cbor_item_t& map,
+                                              int field_number,
+                                              optional<ProtocolVersion>& out) {
+  cbor_item_unique_ptr field = empty_cbor_ptr();
+  StatusCode sc = GetField(map, field_number, field);
+  if (sc == StatusCode::kNotFound) {
+    out.reset();
+    return StatusCode::kOk;
+  } else if (sc != StatusCode::kOk) {
+    return sc;
+  }
+  int version_int;
+  sc = DecodeInt(*field, &version_int);
+  if (sc != StatusCode::kOk) {
+    return sc;
+  }
+  if (version_int == ProtocolVersion::ONE) {
+    // Initially only version 1 is supported.
+    out.emplace(ProtocolVersion::ONE);
+  } else {
+    out.reset();
```

Comment:
Can you add a ProtocolVerson::ADVANCED to be used when a version greater than the known versions is specified. That way the caller can detect it's dealing with an advanced version. SetProtocolVersion would not allow it to be used.

Reply:
Added a TODO, will do in a clean/new CL.